src/tools/rbd_mirror/instance_watcher/Types.h

```diff
@@ -39,8 +39,110 @@ WRITE_CLASS_ENCODER(PeerImageId);
 
 typedef std::set<PeerImageId> PeerImageIds;
 
+enum NotifyOp {
+  NOTIFY_OP_IMAGE_ACQUIRE  = 0,
+  NOTIFY_OP_IMAGE_RELEASE  = 1,
+};
+
+struct ImageAcquirePayload {
+  static const NotifyOp NOTIFY_OP = NOTIFY_OP_IMAGE_ACQUIRE;
+
+  uint64_t request_id;
+  std::string global_image_id;
+  PeerImageIds peers;
```

Comment:
If it is `ImageReplayer` who should scan all peers, then I suppose `ImageReplayer::add_remote_image/remove_remote_image` methods should retire too? You added them recently so I thought you did not plan this would be the `ImageReplayer` job. Then we need to add  `ImageReplayer::add_peer/remove_peer` instead?

Talking about the race you described, it does not look like hurt and will resolve eventually. Anyway, the `ImageReplayer` looks like will need to recover from such situations to choose a proper peer image. Still the images themselves could be provided externally.

I am ok to do this either way, just want to make it clear.